_overviews/overview/a-taste-of-scala.md

```diff
@@ -0,0 +1,1329 @@
+
+val p = Person("Julia", "Kern")
+println(p.firstName)   // "Julia"
+p.lastName = "Manes"
+p.printFullName()      // "Julia Manes"
+```
+
+
+### FP data modeling 
+
+When writing code in an FP style, you’ll use these constructs:
+
+- Enums to define ADTs
+- Case classes
+- Traits
+
+
+#### Enums
+<!-- 
+- QUESTION: when to use enum vs case class?
+- http://dotty.epfl.ch/docs/reference/enums/enums.html 
+- http://dotty.epfl.ch/docs/reference/enums/adts.html
+- enums can have parameters
+- enums can have methods
+- represented as sealed classes that extend the scala.Enum trait
+- make compatible with java by extending java.lang.Enum
+-->
+
+The `enum` construct is a great way to model algebraic data types (ADTs) in Scala 3. For instance, a pizza has three main attributes:
+
+- Crust size
+- Crust type
+- Toppings
+
+These are concisely modeled with enums:
+
+```scala
+enum CrustSize:
+  case Small, Medium, Large
+
+enum CrustType:
+  case Thin, Thick, Regular
+
+enum Topping:
+  case Cheese, Pepperoni, BlackOlives, GreenOlives, Onions
+```
+
+Once you have an enum you can use it in all of the ways you normally use a trait, class, or object:
+
+```scala
+import CrustSize._
+val currentCrustSize = Small
+
+// enums in a `match` expression
+currentCrustSize match 
+  case Small => println("Small crust size")
+  case Medium => println("Medium crust size")
+  case Large => println("Large crust size")
+
+// enums in an `if` statement
+if currentCrustSize == Small then println("Small crust size")
+```
+
+Enums can also take parameters and have user-defined members like fields and methods. Here’s a sneak-peek of those capabilities:
+
+```scala
+enum Planet(mass: Double, radius: Double):
+  private final val G = 6.67300E-11
+  def surfaceGravity = G * mass / (radius * radius)
+
+  case Mercury extends Planet(3.303e+23, 2.4397e6)
+  case Venus   extends Planet(4.869e+24, 6.0518e6)
+  // more ...
+}
+```
+
+Enums are covered in detail in the Data Modeling section of this Overview, and in the Reference documentation.
+
+
+#### Case classes
+
+A *case class* is an extension of the base Scala class. Case classes provide features that make them useful for functional programming. They have all of the functionality of a regular class, and more. When the compiler sees the `case` keyword in front of a `class`, it generates code for you, with these benefits:
+
+- Case class constructor parameters are public `val` fields by default, so accessor methods are generated for each parameter.
+- An `apply` method is created in the companion object of the class, so you don’t need to use the `new` keyword to create a new instance of the class.
+- An `unapply` method is generated, which lets you use case classes in more ways in `match` expressions.
+- A `copy` method is generated in the class. This provides a way to clone an object while making updates to its values during the cloning process.
+- `equals` and `hashCode` methods are generated.
```

Comment:
I think this is the most important difference between classes and case classes, and for me this is often the criterion that determines whether I need a class or a case class. I believe we should go further and explain what does the equality mean on case classes vs regular classes.
Or maybe we do it later in the overview?

Reply:
That’s a good point, thanks! I think that sounds like a detail I should refer readers to later in the overview. I’ll add something here.

I remember reading in *Programming in Scala*, that “The biggest advantage of case classes is that they support pattern matching.” I debated about including that here, and then I thought it was a detail I should discuss later.